crypto/src/main/java/org/cryptimeleon/incentive/crypto/Setup.java

```diff
@@ -147,22 +141,13 @@ public static ProviderKeyPair providerKeyGen(IncentivePublicParameters pp) {
         // generate PRF key for provider
         PrfKey betaProv = pp.getPrfToZn().generateKey();
 
-        // generate SPS-EQ key pair
-        SignatureKeyPair<SPSEQVerificationKey, SPSEQSigningKey> spsEqKeyPair = pp.getSpsEq().generateKeyPair(3);
-        SPSEQVerificationKey spseqVerificationKey = spsEqKeyPair.getVerificationKey();
-        SPSEQSigningKey spseqSigningKey = spsEqKeyPair.getSigningKey();
+        // generate SPS-EQ key pairs
+        SignatureKeyPair<SPSEQVerificationKey, SPSEQSigningKey> tokenSpsEqKeyPair = pp.getSpsEq().generateKeyPair(3);
+        SignatureKeyPair<SPSEQVerificationKey, SPSEQSigningKey> genesisSpsEqKeyPair = pp.getSpsEq().generateKeyPair(2);
 
         // wrap up values
-        ProviderPublicKey pk = new ProviderPublicKey(spseqVerificationKey, h);
-        ProviderSecretKey sk = new ProviderSecretKey(spseqSigningKey, q, betaProv);
+        ProviderPublicKey pk = new ProviderPublicKey(tokenSpsEqKeyPair.getVerificationKey(), genesisSpsEqKeyPair.getVerificationKey(), h);
+        ProviderSecretKey sk = new ProviderSecretKey(tokenSpsEqKeyPair.getSigningKey(), genesisSpsEqKeyPair.getSigningKey(), q, betaProv);
         return new ProviderKeyPair(sk, pk);
     }
-
-
-    // Enum for choosing the bilinear group
```

Comment:
this explains why the group choice was marked as entirely new up above

Reply:
Yes!